hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerMerkleTree.java

```diff
@@ -0,0 +1,180 @@
+
+    for (BlockMerkleTree blockTree: id2Block.values()) {
+      ContainerProtos.BlockMerkleTree blockTreeProto = blockTree.toProto();
+      containerTreeBuilder.addBlockMerkleTree(blockTreeProto);
+      // Add the block's checksum to the buffer that will be used to calculate the container checksum.
+      containerChecksumBuffer.putLong(blockTreeProto.getBlockChecksum());
+    }
+    containerChecksumBuffer.flip();
+    checksumImpl.update(containerChecksumBuffer);
+
+    return containerTreeBuilder
+        .setDataChecksum(checksumImpl.getValue())
+        .build();
+  }
+
+  /**
+   * Represents a merkle tree for a single block within a container.
+   */
+  private static class BlockMerkleTree {
+    // Map of each offset within the block to its chunk info.
+    // Chunk order in the checksum is determined by their offset.
+    private final SortedMap<Long, ChunkMerkleTree> offset2Chunk;
+    private final long blockID;
+
+    BlockMerkleTree(long blockID) {
+      this.blockID = blockID;
+      this.offset2Chunk = new TreeMap<>();
+    }
+
+    /**
+     * Adds the specified chunks to this block. The offset value of the chunk must be unique within the block,
+     * otherwise it will overwrite the previous value at that offset.
+     *
+     * @param chunks A list of chunks to add to this block.
+     */
+    public void addChunks(Collection<ChunkInfo> chunks) {
+      for (ChunkInfo chunk: chunks) {
+        offset2Chunk.put(chunk.getOffset(), new ChunkMerkleTree(chunk));
+      }
+    }
+
+    /**
+     * Uses chunk hashes to compute a block hash for this tree, and returns it as a protobuf object. All block checksum
+     * computation for the tree happens within this method.
+     *
+     * @return A complete protobuf object representation of this block tree.
+     */
+    public ContainerProtos.BlockMerkleTree toProto() {
+      ContainerProtos.BlockMerkleTree.Builder blockTreeBuilder = ContainerProtos.BlockMerkleTree.newBuilder();
+      ChecksumByteBuffer checksumImpl = ChecksumByteBufferFactory.crc32Impl();
+      ByteBuffer blockChecksumBuffer = ByteBuffer.allocate(Long.BYTES * offset2Chunk.size());
+
+      for (ChunkMerkleTree chunkTree: offset2Chunk.values()) {
+        // Ordering of checksums within a chunk is assumed to be in the order they are written.
+        // This assumption is already built in to the code that reads and writes the values (see
+        // ChunkInputStream#validateChunk for an example on the client read path).
+        // There is no other value we can use to sort these checksums, so we assume the stored proto has them in the
+        // correct order.
+        ContainerProtos.ChunkMerkleTree chunkTreeProto = chunkTree.toProto();
+        blockTreeBuilder.addChunkMerkleTree(chunkTreeProto);
+        blockChecksumBuffer.putLong(chunkTreeProto.getChunkChecksum());
+      }
+      blockChecksumBuffer.flip();
+      checksumImpl.update(blockChecksumBuffer);
+
+      return blockTreeBuilder
```

Comment:
Maybe we just pass the builder to ChecksumTreeManager and let it build it once? Will need to look into the proto methods.

Reply:
Discussed offline, but the goal here is to minimize the possibility of building an invalid tree proto. The approach here requires callers like the scanner to only provide the chunks they encounter, which fits well with its current implementation. In order to get a proto file, callers must use this method, which ensures the whole tree is properly filled in and the logic to do that is in one central spot.